Bullet-dodging game/Bullet-Dodging Game in Processing.md

```diff
@@ -0,0 +1,2191 @@
+  player = new Player();
+  life = new Life();
+  for (int i = 0; i < MAX_BULLET; i++) {
+    bullets[i] = new Bullet(- i * height / MAX_BULLET);
+  }
+}
+
+void draw() {
+  background(255);      // white background
+  player.update();
+  player.show();
+  for (int i = 0; i < MAX_BULLET; i++) {
+    bullets[i].update();
+    bullets[i].show();
+    if (bullets[i].check_collision()) {
+      hp --;
+      bullets[i].ypos = - bullets[i].size;
+      if (hp == 0) {
+        exit();
+      }
+    }
+  }
+  life.show();
+  if (life.check_collision()) {
+    life = new Life();
+  }
+  show_status();
+  score ++;
+}
+
+void show_status() {
+  fill(255);
+  stroke(0);
+  strokeWeight(3);
+  rect(80, 28, 145, 27, 5);
+  image(heart1, 20, 19, 45, 45);
+  fill(255, 0, 0);
+  noStroke();
+  rect(82, 30, 142 * hp / MAX_HP, 24, 3);
+  fill(0);
+  strokeWeight(5);
+  textSize(24);
+  text(hp + " / 6", 125, 50);  
+  text("Score: " + score, 1000, 50);
+}
+```
+
+
+
+## Part IV: Adding Enemies and Shooting
+
+We have already implemented all the essential functions of a simple bullet-dodging game. If you're ready for a more advanced program, it is now time to add enemies to the game and make it a shooting game.
+
+First of all, it is imperative that we declare a global time variable `int time = 0;` and increment it (`time ++;`) at the end of `draw()`, so that we can decide when the enemy will fire the next bullet regularly.
+
+We would like to have different kinds of enemies shooting different kinds of bullets, so we will abandon our previous Bullet class as well as the bullet array. Instead, let's first define the following two abstract classes--Bullet and Enemy--which requires support for the basic functions such as `update()` and `show()`. Each subclass will have its own attributes--size, speed, etc.
+
+```
+abstract class Bullet {
+  int size, speed, damage;
+  float xpos, ypos, xspeed, yspeed;
+
+  // decrement hp and return true if the bullet hits the player
+  boolean check_collision() {
+    if (dist(xpos, ypos, player.xpos, player.ypos) < player.size / 2) {
+      hp -= damage;
+      return true;
+    }
+    return false;
+  }
+
+  abstract boolean update();  // update the position of the bullet
+  abstract void show();  // draw the bullet on the screen
+}
+
+abstract class Enemy {
+  int size, speed, max_hp, hp, appear_time = time;
+  float xpos, ypos;
+  
+  // draw the hp bar above the enemy
+  void show_hp() {
+    fill(255);
+    stroke(0);
+    strokeWeight(2);
+    rect(xpos - 60, ypos - 100, 120, 10);  // total hit points
+    noStroke();
+    fill(255, 0, 0);
+    rect(xpos - 59, ypos - 99, 119 * hp / max_hp, 8); // remaining hit points
+  }
+
+  abstract void update();  // update the position of the bullet
+  abstract void show();    // draw the enemy on the screen
+  abstract void shoot();   // fire a bullet
+}
+```
+
+Abstract classes cannot be instantiated, we are able to create arrays holding objects of their subclasses:

```

Comment:
these are lists, not arrays, one of the code or the descriptions need to be updates